src/v/kafka/server/handlers/fetch.cc

```diff
@@ -545,6 +545,10 @@ static void fill_fetch_responses(
             octx.rctx.probe().record_fetch_latency(fetch_latency);
         }
     }
+
+    if (likely(total_memory_units.has_value())) {
+        octx.adopt_fetch_memory_units(std::move(total_memory_units).value());
```

Comment:
Things get a little tricky here. `fill_fetch_responses` can be called multiple times for a given partition. Each time its called it overrides the the previous response for the partition with a new one. So we don't want to accumulate the semaphore units for all the times its called. But rather only keep the units from the last time its called for a given partition and release the rest.

This likely means that we need to track per-partition semaphore units unfortunately. So that when a response to a given partition is overridden we release the old units and hold on the the units for the new response instead.